backend/src/process.py

```diff
@@ -168,6 +168,37 @@ async def run(
                 "Errors occurred during iteration: \n• {}".format("\n• ".join(errors))
             )
 
+    async def run_while(
+        self,
+        length: int,
+        before: Callable[[int], Union[None, Literal[False]]],
+    ):
+        errors: List[str] = []
+        index = 0
+        while True:
+            try:
+                await self.progress.suspend()
+
+                result = before(index)
+                if result is False:
+                    break
+
+                await self.run_iteration(index, length)
+
+                if index < length:
+                    index += 1
```

Comment:
Right now, progress is not correctly calculated if the number of actual iterations is different from `length`.

If there are more iterations, then `index` will be equal to `length` and since `run_iteration` sends out `(index+1)/length` as progress, this will result in progress >100%. This is not a problem with `run_iteration` since that function assumes valid indexes (as you would).

If there are fewer iterations, then the progress never reaches 100%. This should be easy to fix by sending out 100% at the end of the while loop if `index < length`.

Also, the current `index += 1` doesn't run if either `before(index)` or `run_iteration` errors. This is a huge problem because `before` might require the index to be correct.